WordPress/gutenberg
# [Mobile] - Drag & drop blocks - Fetch and share blocks layout size and position coordinates

Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4601

`Gutenberg Mobile PR` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/4611

## Description
This PR includes the following changes:

**Block List context was moved into its own file to expand its functionality:**

- Changes the default value of the context to include: 
  - `scrollRef` (before it was on the top level).
  - `blocksLayouts` to store the blocks `onLayout` data.
  - `updateBlocksLayouts` to be able to update the `blocksLayouts` data.
- Adds `useBlockListContext` to be able to access the context.
- `findByRootId` to look for a block's data by root id.
- `deleteByClientId` to delete a block's data from `blocksLayouts`.

**`BlockListItemCell` new component:**

This is now a wrapper for every item of the block list FlatList, it will use `onLayout` to store the block's size and coordinates into the Block list context: `blocksLayouts`. This component is set using `CellRendererComponent` this is needed in order to get the correct `x`, `y` coordinates. Trying to get this data within `renderItem` would return `0` for the `x` and `y` coordinates.

**An example of the data stored in `blocksLayouts`:**

<details><summary>Expand</summary>
<p>

```json
{
    "c8965391-4798-4ba5-b3b3-e51be1f84a33": {
        "clientId": "c8965391-4798-4ba5-b3b3-e51be1f84a33",
        "width": 390,
        "height": 241.33334350585938,
        "x": 0,
        "y": 164,
        "innerBlocks": {
            "b6334622-c13c-4a9b-bafc-5e467592b1b2": {
                "clientId": "b6334622-c13c-4a9b-bafc-5e467592b1b2",
                "rootClientId": "c8965391-4798-4ba5-b3b3-e51be1f84a33",
                "width": 187,
                "height": 209.33334350585938,
                "x": 187,
                "y": 0,
                "innerBlocks": {}
            },
            "054e6463-1c6c-49fe-8d92-92cd2259ee4b": {
                "clientId": "054e6463-1c6c-49fe-8d92-92cd2259ee4b",
                "rootClientId": "c8965391-4798-4ba5-b3b3-e51be1f84a33",
                "width": 187,
                "height": 209.33334350585938,
                "x": 0,
                "y": 0,
                "innerBlocks": {}
            }
        }
    },
    "ad9b6a24-09ae-48c8-87a1-93976485b0ee": {
        "clientId": "ad9b6a24-09ae-48c8-87a1-93976485b0ee",
        "width": 390,
        "height": 81,
        "x": 0,
        "y": 83,
        "innerBlocks": {}
    },
    "f860cafe-d9f2-4916-a161-d96855f9c06c": {
        "clientId": "f860cafe-d9f2-4916-a161-d96855f9c06c",
        "width": 390,
        "height": 56.666656494140625,
        "x": 0,
        "y": 405.3333435058594,
        "innerBlocks": {}
    }
}
```

</p>
</details>

## Testing Instructions

### Manual testing

- Apply the following patch:

```patch
diff --git a/packages/block-editor/src/components/block-list/block-list-context.native.js b/packages/block-editor/src/components/block-list/block-list-context.native.js
index dc160a814..bfef33f67 100644
--- a/packages/block-editor/src/components/block-list/block-list-context.native.js
+++ b/packages/block-editor/src/components/block-list/block-list-context.native.js
@@ -82,6 +82,8 @@ function updateBlocksLayouts( blockData ) {
 			};
 		}
 	}
+
+	console.log( JSON.stringify( blocksLayouts.current ) );
 }
 
 export {
```

- Open the editor
- Add some blocks, including inner blocks
- **Expect** to see the output data `console.log` to include the blocks
- Remove a block
- **Expect** to see that block removed from the output `console.log`

## Screenshots 
N/A

## Types of changes
New feature

## Checklist:
